### DISCUSSION TITLE
Coap gateway POST - bad request

### DISCUSSION THREAD

**Discussion by Michaelcombs** [emqx/emqx#11772] [Q&A] [answered]
I'm using the COAP AT commands in the SIM7000 modem to communicate with the EMQX 5.2 Coap Gateway. The gatreway is configured to 'require connection' and I can successfully connect and see my coap client in the EMQX coap web interface.

When I attempt to POST a message to the gateway to be published via mqtt, I receive a bad request response (4.00) with no payload information indicating the problem. 

Here is my AT command. You can see the uri and querystring containing my clientid and token. The payload is "data" and the message id is 2.

`AT+CCOAPPARA=code,2,uri-path,0,"mqtt/ps/test/1103/data/net",uri-query,0,"clientid=PS1103&token=2151045968&retain=true&qos=0&expiry=0",payload,0,"data",mid,2,type,"CON"`

I have enabled debug logging in EMQX. Here is the log entry associated with the POST:

`2023-10-16T23:55:44.268121+00:00 [debug] msg: RECV_packet, mfa: emqx_gateway_conn:handle_incoming/2, line: 750, peername: 172.18.0.2:38765, packet: {coap_message,con,post,3,<<>>, #{uri_path =>, [<<"mqtt">>,<<"ps">>,<<"test">>,<<"1103">>,<<"data">>, <<"gps">>], uri_query =>, #{<<"clientid">> => <<"PS1103">>,<<"expiry">> => <<"0">>, <<"qos">> => <<"0">>,<<"retain">> => <<"false">>, <<"token">> => <<"3261410743">>}}, <<"data">>}`

`2023-10-16T23:55:44.268346+00:00 [debug] msg: SEND_packet, mfa: emqx_gateway_conn:serialize_and_inc_stats_fun/1, line: 819, peername: 172.18.0.2:38765, packet: {coap_message,ack,{error,bad_request},3,<<>>,#{},<<>>}`

`2023-10-16T23:55:44.268460+00:00 [debug] msg: SEND_data, mfa: emqx_gateway_conn:send/2, line: 854, peername: 172.18.0.2:38765, data: <<96,128,0,3>>`

Please advise.

**Comment by hjianbo** [comment]
Thanks for such detailed feedback. It looks like the wrong URI path was specified when performing the message posting because the message posting URI was prefixed with `/ps` instead of `/mqtt/ps`.

You can refer to [reference-coap-client-guide](https://docs.emqx.com/en/enterprise/latest/gateway/coap.html#reference-coap-client-guide)

**Reply by Michaelcombs** [reply]
The message URI  I used was  "mqtt/ps/test/1103/data/net".  Should I use 'ps/test/1103/data/net'? 

The reference guide is a little confusing:


Request Parameters:

Method: POST

URI: **mqtt/{+topic}**{?QueryString*}

{+topic} is the topic of publish messages, i.e. the URI is **ps/coap/test** if publish message to coap/test.

**Reply by hjianbo** [reply, marked as answer]
Sorry, there is an error in the description of the document here. I have fix it via https://github.com/emqx/emqx-docs/pull/2178

But your understanding is correct, i.e: **the URI is ps/coap/test if publish message to coap/test.**
